filament/backend/include/backend/Platform.h

```diff
@@ -66,6 +66,8 @@ class UTILS_PUBLIC Platform {
      * thread, or if the platform does not need to perform any special processing.
      */
     virtual bool pumpEvents() noexcept { return false; }
+    
+    virtual void* getMTLDevice() noexcept { return nullptr; }
```

Comment:
no we can't have a metal specific API on Platform. This would have to go on MetalPlatform only.

Reply:
MetalPlatform.h is a private header. We can't access it unless making it public.

Reply:
Could you describe your use case in more details then? We keep the backends private for good reasons.

Reply:
There are multiple rendering framework in our video processing app, snapchat/TikTok for example. The whole frame is consist of 2D renderer(such as GPUImage), 3D renderer(such as filament) and then 2D renderer(such as GPUImage). It's good to keep the same physical devices such MTLDevice and EGLContext between them. Although filament provide Platform creation, we just can't use those gpu hardware it created. Hence here's a proposal to make it accessible.

Reply:
We don't think "it's good" to share contextes.  In particular sharing the opengl context means your app (or other libraries) can screw-up our context, leading to impossible to debug bugs. There are also threading implications.

I still don't understand what you cannot do with the current system.

Reply:
@Michael-Lfx assume Filament always simply called `MTLCreateSystemDefaultDevice`—if you also called `MTLCreateSystemDefaultDevice`, would that solve your problems (at least for Metal)?

Reply:
> We don't think "it's good" to share contextes. In particular sharing the opengl context means your app (or other libraries) can screw-up our context, leading to impossible to debug bugs. There are also threading implications.
> 
> I still don't understand what you cannot do with the current system.

How can we share GPU texture between different rendering frameworks which owns their own E(A)GLContext? ReadPixels back to CPU and then upload to GPU to pass data between them will greatly impact performance. 

Filament support `static Engine* create(Backend backend = Backend::DEFAULT, Platform* platform = nullptr, void* sharedGLContext = nullptr)` to make sure resources created inside filament are visible to caller. Say we Texture::import(a_gl_tex_id) to be a sample source or color attachment for framebuffer.

Same situation for Metal and Vulkan. Using the same MTLDevice/VkPhysical/VkDevice to create GPU resources are visible for other rendering frameworks to sample or draw to filament created textures/buffers.

Reply:
The support for a shared EGL context is something we did reluctantly for legacy code at Google. The goal was never to make Filament resources available externally, but the other way around. And even then, there are ways to share texture data efficiently via other mechanisms that don't require context sharing.

Reply:
> @Michael-Lfx assume Filament always simply called `MTLCreateSystemDefaultDevice`—if you also called `MTLCreateSystemDefaultDevice`, would that solve your problems (at least for Metal)?

Yes. Calling `MTLCreateSystemDefaultDevice ` both on iOS and macOS can solve my problem. How to suit the Android Vulkan case? Our app is running on macOS, iOS, Android and Windows.

Another plan is changing 
`static Engine* create(Backend backend = Backend::DEFAULT, Platform* platform = nullptr, void* sharedGLContext = nullptr)`
to
```
struct EngineCreateInfo {
    Backend backend = Backend::DEFAULT;
    Platform* platform = nullptr;
    void* sharedGLContext = nullptr;
    // For Metal
    void* mtlDevice = nullptr;
    // For Vulkan (maybe missing VkPhysicalDevice/VkInstance)
    void* vkDevice = nullptr;
};
static Engine* create(EngineCreateInfo &createInfo);
```

Reply:
*If* the goal was to extend sharing to Vulkan and Metal, you could repurpose the existing pointer parameter, there's no need for an entire structure. As for sharing on Android, you should be rendering into `Surface` objects that come from a `SurfaceTexture` or an `ImageReader` and use that. There shouldn't be a reason to share the contexts.

Reply:
Sounds like `sharedGLContext` of `static Engine* create(Backend backend = Backend::DEFAULT, Platform* platform = nullptr, void* sharedGLContext = nullptr)` should be interpret by backends. It can be cast from `void*` to `id<MTLDevice>` inside `Driver* MetalPlatform::createDriver(void* sharedContext) noexcept`.

Reply:
I really would like to move this discussion to a proper discussion on GitHub so it can be useful to other people.

We should discuss this before we implement solutions that come with their own problems.

I'm going to close this PR. Again let's discuss the problem, and figure out what existing facilities you can use.